### ISSUE TITLE
Support for advanced filesystem features (setuid/setgid/xattrr/hardlinks/sockets, etc)

### ISSUE THREAD

**Issue by mbiebl** [kopia/kopia#544]
Version: 0.6.1

Followup to #543 

```
# ls -al /usr/bin/sudo
-rwsr-xr-x 1 root root 147560 Feb  2  2020 /usr/bin/sudo
```
After the `kopia restore`
```
# ls -al /var/tmp/restore/sudo
-rwxr-xr-x 1 root root 147560 Feb  2  2020 /var/tmp/restore/sudo
```

I assume aside from the SUID bit, SGID and sticky bit are also not (yet) properly supported.

**Comment by jkowalski** [comment]
Yes, there's a whole bunch of things currently not captured at the filesystem level, including setuid/gid, hardlinks, mount points, sockets, xattr, ACLs, etc.

Currently only files (simple, without alternate streams) and symlinks are supported, and only simple filemode + owner/group are supported.

I'm going to repurpose this issue to track adding the missing pieces and close the others as duplicates.

**Comment by mbiebl** [comment]
Sounds fine to me, thanks. Wasn't sure if you preferred individual issues or not.

**Comment by mbiebl** [comment]
> Currently only files (simple, without alternate streams) and symlinks are supported

Hm, see #543. When running `kopia restore`, no symlinks were restored. Browsing the repository via `kopia-ui` though, I see those files.

**Comment by jolly-jim** [comment]
Ideally this will include - under Windows - support for: 
- File attributes
- ACLs
- Symlinks (file & folder), Junction points & Hard links.

IMO, for the latter, the required behaviour should be to just capture the symlink/junction point/hard link itself, and **not** it's target.
If there really is a need to capture the target (or more accurately the target's content), then that could be handled with an optional - non default - argument.

**Comment by WMP** [comment]
If you write this feature kopia will be good tool to backup whole VM, now i can do this with xfs_dump tool.

**Comment by dima-altinity** [comment]
@jkowalski 
very interesting project and I also think that this feature is really missing.
whats the current status of symlinks and hardlinks support?

**Comment by jkowalski** [comment]
Symlinks and hardlinks are correctly backed up. Both can be restored but hardlinks are restored as individual files and not as hardlinks.

**Comment by dima-altinity** [comment]
thnx @jkowalski , just would like to clarify how does it work( restore) if there are mutliply hard links to the same file?
lets say I have file  /home/user/1.tar.gz ( size 1Gb) and /home/user/link1.gz, /home/user/link2.gz  that are hard links to /home/user/1.tar.gz
Scenarios
1. I create and upload backup to S3 for /home/user/  ( how much data will be sent and stored on S3 -  1Gb or 3Gb ?)
2. I create /home/user/link3.gz hardlink  that points to the same file /home/user/1.tar.gz. Then upload backup to S3. How much data will sent to S3 during this operation?
3. I restore backup from S3 to local ( backup contains 1 file and 3 hard links). How much data will be downloaded from S3? 1Gb or 4Gb?
4. do I understand right that after restore this backup I'll get 4 files in /home/user/ with total size 4Gb ?

**Comment by jkowalski** [comment]
1. 1gb because of deduplication 
2. 0 (plus minimal overhead for indices and manifests)
3. Physically 1 gb will be downloaded (because of caching) but logically each copy will be independently fetched (so 3gb). 
4. That is correct.

**Comment by dima-altinity** [comment]
how much efforts is required to add full support of hard links ( so that #3 and #4 worked properly)? 
is it just changes in code? or changes in architecture also required  ?

**Comment by jkowalski** [comment]
It's mostly some additional restore work, architecture is fine.

**Comment by martinpk** [comment]
Sockets: until full support is implemented, would you reword this message:

`! Ignored error when processing "<filename>": unknown or unsupported entry type`

to indicate that (1) the file in question is a socket, and (2) it has not been backed up?

Also in the interim, would you provide a non-default option to ignore sockets, in order to reduce noise?

**Comment by dima-altinity** [comment]
@jkowalski 
does this latest push add full support of hard links? ( backup & restore)

**Comment by jkowalski** [comment]
No it does not, will reopen.

**Comment by yacoob** [comment]
One extra comment: it'd be worthwhile to add this limitation to the documentation, to make sure people are aware of this limitation, instead of discovering it at the test - or worse, at restore time.

**Comment by seaniedan** [comment]
I'm currently snapshotting a disk full of backups made with the excellent [RaspiBackup](https://github.com/framps/raspiBackup). It has a folder structure of incremental backups using hardlinks. It is taking a **very** long time to do the first snapshot! 

Would it be quicker if I snapshot the latest backup first? Then add the whole disk?

**Comment by seaniedan** [comment]
In case it's useful for discussion or when writing documentation, [this forum page on Duplicati](https://forum.duplicati.com/t/hardlinks-handling/3769) has some good explanations of the difficulties around hardlinks.

**Comment by seaniedan** [comment]
Hi Jarek, you say supporting hardlinks is "mostly some additional restore work" - Kopia backups with no hardlinks take ~11 seconds for 2.5 TB of files, but a much smaller directory of hardlinks takes ~9 hours. Nothing seems to bring this time down. Is this time expected to come down with an update?

**Comment by aniqueta** [comment]
Still needed for macOS users. Thanks!

**Comment by xeor** [comment]
I don't think the fix from @PhracturedBlue is merged into kopia yet? Still a wip maybe?

**Comment by PhracturedBlue** [comment]
Something screwy going on in github.  I don't think I ever touched this code, it somehow got attributed to be due to an unrelated PR.

The fix in question seems to be this one: https://github.com/kopia/kopia/pull/2597 which was merged, though I don't think it entirely fixes the issue (based on description).

**Comment by miekg** [comment]
Hi all, I've found myself in need for support for xattr (on Linux) and was thinking along the following lines, put here to see if this is something I should undertake.

Restic is also in Go and matches license wise and *does* support xattr. Can I, should I ?, lift and modify that code and prepare a PR here?

**Comment by miekg** [comment]
this has become: https://github.com/kopia/kopia/pull/3643

**Comment by guiltlab** [comment]
If I understand correctly, kopia is able to backup a file structure containing hardlinks correctly, but not restore it with preserved hardlinks? Is this still the current situation?

It should be workable for me even if not ideal. It looks like we can mount a Kopia snapshot to a folder. If the hardlinks are preserved in the snapshot, then they may be able to be restored with another tool which handles that correctly such as FastCopy.

**Comment by l-salewski** [comment]
How does kopia handle hardlinks during the backup, if a file is hardlinked to a different location that is also part of the backup? Does Kopia read the file once, realize its other location is also part of the backup and store that? Or does it read it multiple times from disk, only for the de-duplication to realize that this file has already been saved? The former would be much quicker of course.

Edit: I found https://github.com/kopia/kopia/pull/4460, which states that it is currently the latter.
